File: src/main/java/com/warren/bt/RandomSelector.java
package com.warren.bt;

import java.util.Random;

public class RandomSelector extends Composite {
  private final Random random = new Random();
  private Node selectedChild = null;

  @Override
  public Status tick() {
    if (!hasChildren()) {
      return Status.FAILURE;
    }

    if (selectedChild == null) {
      int index = random.nextInt(children.size());
      selectedChild = children.get(index);
    }

    Status status = selectedChild.tick();

    if (status == Status.SUCCESS || status == Status.FAILURE) {
      reset();
    }

    return status;
  }

  @Override
  public void reset() {
    super.reset();
    selectedChild = null;
  }
}

File: src/main/kotlin/com/warren/bt/RandomSelector.kt
package com.warren.bt

class RandomSelector : Composite("RandomSelector") {
  private var order: MutableList<Int> = mutableListOf()
  private var idx = 0

  override fun tick(): Status {
    if (order.isEmpty()) {
      order = children.indices.shuffled().toMutableList()
      idx = 0
    }
    while (idx < order.size) {
      val status = children[order[idx]].tick()
      when (status) {
        Status.SUCCESS -> {
          return Status.SUCCESS
        }
        Status.FAILURE -> idx++
        Status.RUNNING -> {
          return Status.RUNNING
        }
      }
    }
    return Status.FAILURE
  }

  override fun onReset() {
    order.clear()
    idx = 0
  }
}
